Added sysid and pathweaver testing, fixed mistakes:

SusanC3 opened the pull request:
* Added new sysid config file
* Added test path that goes 3 meters in x direction
* Added PID + feedforward functionality & sdb outputs for swerve module drive motors
* Added necessary initialization for the ramsete command in initializeAutoTraj
* clamps/volt values in differentialDrive are now done as doubles not volts to prevent errors
* note: if x and y tangents are 0, curvature can be null & feedforward will break

SusanC3 committed:
Added sysid and pathweaver testing, fixed mistakes:
* Added new sysid config file
* Added test path that goes 3 meters in x direction
* Added PID + feedforward functionality & sdb outputs for swerve module drive motors
* Added necessary initialization for the ramsete command in initializeAutoTraj
* clamps/volt values in differentialDrive are now done as doubles not volts to prevent errors
* note: if x and y tangents are 0, curvature can be null & feedforward will break

gaborszita commented on .vscode/settings.json:
```diff
@@ -83,6 +83,26 @@
     "xloctime": "cpp",
     "xtree": "cpp",
-    "*.inc": "cpp"
+    "*.inc": "cpp",
```
I'm not sure what this config does or why it is necessary. Personally, I would gitignore everything in the `.vscode` directory, but this is just personal preference. I'm not saying we should delete the whole `.vscode` directory and gitignore it, I just want to say my opinions on this.  So far I found it best to gitignore `.idea` and `.vscode` folders since it changes if a developer changes an IDE setting.

What I would like to say with this is that you should check if this is a justified change or if you just accidentally committed this. I have some changes in my `settings.json` already and I don't want to unnecessarily resolve merge conflicts. But again, don't get me wrong - if this is a justified change, go ahead and keep it. It just seems to me that this was committed by mistake since you incorporated it in a larger commit of a completely different thing.

SusanC3 replied:
So far I've just been using the default `.gitignore` file provided by WPILib. I agree these merge conflicts are annoying; I'll have it ignore everything in the `.vscode` folder (we don't use `.idea`) and if it creates problems we can change it back.

gaborszita replied:
Where did you get this `.gitignore` from? It's a bit weird because some files in `.vscode` are ignored and some aren't:
```
.vscode/*
!.vscode/settings.json
!.vscode/tasks.json
!.vscode/launch.json
!.vscode/extensions.json
```
The `.gitignore` of wpilib here ignores the whole `.vscode` directory: [https://github.com/wpilibsuite/vscode-wpilib/blob/main/.gitignore](https://github.com/wpilibsuite/vscode-wpilib/blob/main/.gitignore)

SusanC3 replied:
It's auto generated when a new WPILib vscode project is made. Pretty sure this is the official file: https://github.com/wpilibsuite/vscode-wpilib/blob/main/vscode-wpilib/resources/gradle/shared/.gitignore

gaborszita replied:
Okay, this seems fine. I'll create a better way to solve this and create PR for it this week.

gaborszita commented on src/main/deploy/paths/3mTest.wpilib.json:
```diff
@@ -0,0 +1,587 @@
+[
```
Can you please add some explanation on what this file is? Is this for PathWeaver? I don't really understand what this file is, and we already have a lack of documentation, so I think it would be important to add some explanation to what this file is. It doesn't have to be in this file (it also can't since JSON doesn't have comments), so I think you should create a README in the `paths` directory that explains what the files here are.

SusanC3 replied:
This is for SysId; it is a config file that sets up the drive motors for the 3 connected swerve modules on the extra swerve base, to be loaded into the SysId application. I will add a Readme both for the sysid and the paths folder.

I agree that especially the higher level repository structure (that is harder to document through code comments) needs better documentation. I made a [taskade](https://www.taskade.com/d/fao3qGrbzxfXddXG) assignment for documentation, aside from comments in the code what do you think we should use?

gaborszita commented on src/main/include/Robot.h:
```diff
@@ -35,4 +35,5 @@ class Robot : public frc::TimedRobot {
    //subject to adjustment
   void joy_val_to_mps(double& val) { val *= 4; }
```
Not a huge problem, and I know this is not part of this commit, I would just like to share my opinions on this. Why are you passing the double argument as a reference? I know the reason is that you are modifying the value, but then why not just reassign it where it is called, like this:
`valToSet = joy_val_to_mps(valToSet);`
Again, this is not a huge problem, I just think it would be much simpler and more readable if you reassigned it rather than passing in the value as reference. There is no (or extremely little) performance changes doing this.

SusanC3 replied:
There's no real reason; I initially wrote those functions when I was working with a programmer (outside of robotics) who loved modifying values like this and it rubbed off on me. This implementation is messy overall, I'll change it to re-assign the value and I'll put the definition in Robot.cpp (especially since we may modify or complicate the function to better match the speed profile of the old/offseason robot).

gaborszita reviewed (changes requested):
Please see my comments.

gaborszita committed:
Fix CodeFactor issue in PR

The CodeFactor issue in the PR #2 is fixed by this commit. The redundant new line is removed.

SusanC3 committed:
Addressed comments:
* Commented out .vscode folder includes in .gitignore
* Added READMEs to the PathWeaver and SysId folder
* Changed joy_val_to_mps and joy_val_to_rps to return value, implemented in Robot.cpp
* Limelight lights turned off
* Fixed some indentation

SusanC3 reviewed (commented):
I have addressed the comments.

SusanC3 reviewed (commented):
(Can I not delete my review comments?)

SusanC3 committed:
fixed mistake: didn't make conversion functions members of Robot

gaborszita commented:
They look mostly good to me, but I don't know what effect removing those files from `.vscode` will have. I'll check it out later today if it does something bad if we remove it.

gaborszita reviewed (approved):
Great!

linc-mvla commented on src/main/cpp/Robot.cpp:
```diff
@@ -60,6 +63,22 @@ void Robot::SimulationInit() {}
+ * TODO: change this to match the angular velocity calculations from 2022-Offseason
+**/
+double Robot::joy_rot_to_rps(double rot) {
+  return rot * 3*3*M_PI;
```
Is this in radians per second or rotations per second? A max of 27 (rotations) or 4.5 (radians) rpm seems a bit high. random 3s are confusing.

SusanC3 replied:
I actually don't remember how I came up with this. I'll try to change the joystick to mps and rps functions to better match what the calculations would be in 2022-Offseason & i'll improve the clarity/documentation

SusanC3 replied:
In 2022-Offseason, joystick inputs are converted directly to volts (-1 to 1 range mapped to -12 to 12). To match this using velocity velocity, we'll need to know what the speed of the robot would be if the voltage applied to the drive motors is 12V. Until then (and because I doubt we'd ever want to go that fast anyway) I'll put guesses for linear and angular max speeds in the constants file.

linc-mvla commented on PathWeaver/Paths/3mTest.path:
```diff
@@ -0,0 +1,3 @@
+X,Y,Tangent X,Tangent Y,Fixed Theta,Reversed,Name
+0.0,-8.2296,1.0,0.0,true,false,
```
Why is y -8.2296

SusanC3 replied:
Was just an artifact of where I visually created it on the field. It will be changed to 0.0 in the next commit

linc-mvla commented on src/main/cpp/Robot.cpp:
```diff
@@ -34,9 +36,10 @@ void Robot::TeleopPeriodic() {
-  joy_val_to_mps(dx);
-  joy_val_to_mps(dy);
-  joy_rot_to_rps(dtheta);
+  dx = joy_val_to_mps(dx);
```
i think naming it vx might be more intuitive

SusanC3 committed:
Addressed comments:
* made temporary improvement to the joystick to speed mappings
* changed strange -8.2296 y value in pathweaver to 0
* changed dx, dy, dtheta to vx, vy, vtheta

linc-mvla reviewed (commented):
Looks good!
